elrond-wasm/src/types/managed/managed_multi_result_vec.rs

```diff
@@ -0,0 +1,168 @@
+use super::{ManagedBuffer, ManagedType, ManagedVec, ManagedVecItem};
+use crate::{
+    abi::{TypeAbi, TypeDescriptionContainer},
+    api::{EndpointFinishApi, ErrorApi, ManagedTypeApi},
+    contract_base::ManagedSerializer,
+    types::{ManagedArgBuffer, MultiResultVec},
+    ArgId, ContractCallArg, DynArg, DynArgInput, DynArgOutput, EndpointResult,
+};
+use alloc::string::String;
+use core::marker::PhantomData;
+use elrond_codec::{TopDecode, TopEncode};
+
+#[derive(Clone)]
+pub struct ManagedMultiResultVec<M, T>
+where
+    M: ManagedTypeApi,
+{
+    pub(super) raw_buffers: ManagedVec<M, ManagedBuffer<M>>,
+    _phantom: PhantomData<T>,
+}
+
+impl<M, T> ManagedMultiResultVec<M, T>
+where
+    M: ManagedTypeApi,
+{
+    pub fn new_empty(api: M) -> Self {
+        ManagedMultiResultVec {
+            raw_buffers: ManagedVec::new_empty(api),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<M, T> ManagedMultiResultVec<M, T>
+where
+    M: ManagedTypeApi + ErrorApi,
+    T: TopEncode,
+{
+    pub fn push(&mut self, item: T) {
+        let serializer = ManagedSerializer::new(self.raw_buffers.type_manager());
+        self.raw_buffers
+            .push(serializer.top_encode_to_managed_buffer(&item));
+    }
+}
+
+impl<M> From<ManagedVec<M, ManagedBuffer<M>>> for ManagedMultiResultVec<M, ManagedBuffer<M>>
+where
+    M: ManagedTypeApi,
+{
+    #[inline]
+    fn from(v: ManagedVec<M, ManagedBuffer<M>>) -> Self {
+        ManagedMultiResultVec {
+            raw_buffers: v,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<M, T> ManagedMultiResultVec<M, T>
+where
+    M: ManagedTypeApi,
+{
+    pub fn to_arg_buffer(&self) -> ManagedArgBuffer<M> {
+        ManagedArgBuffer::from_raw_handle(
+            self.raw_buffers.type_manager(),
+            self.raw_buffers.get_raw_handle(),
+        )
+    }
+}
+
+impl<M> ManagedMultiResultVec<M, ManagedBuffer<M>>
+where
+    M: ManagedTypeApi,
+{
+    pub fn into_vec_of_buffers(self) -> ManagedVec<M, ManagedBuffer<M>> {
+        self.raw_buffers
+    }
+}
+
+impl<M, T> ManagedMultiResultVec<M, T>
+where
+    M: ManagedTypeApi + ErrorApi,
+    T: ManagedVecItem<M> + TopDecode,
+{
+    pub fn to_vec(&self) -> ManagedVec<M, T> {
+        let mut result = ManagedVec::new_empty(self.raw_buffers.type_manager());
+        let serializer = ManagedSerializer::new(self.raw_buffers.type_manager());
+        for item in self.raw_buffers.into_iter() {
+            result.push(serializer.top_decode_from_managed_buffer(&item));
+        }
+        result
+    }
```

Comment:
For consistency, methods with the naming `to_other_type` should consume `self`:
```suggestion
    pub fn to_vec(self) -> ManagedVec<M, T> {
        let mut result = ManagedVec::new_empty(self.raw_buffers.type_manager());
        let serializer = ManagedSerializer::new(self.raw_buffers.type_manager());
        for item in self.raw_buffers.into_iter() {
            result.push(serializer.top_decode_from_managed_buffer(&item));
        }
        result
    }
```

Maybe also add some rust-docs explaining that this is an expensive operation. It's not immediately obvious, as in the previous version it was a NOP.

Reply:
"into_..." consumes type,, "to_..." does not